core/common/src/main/java/alluxio/collections/NonUniqueFieldIndex.java

```diff
@@ -38,32 +38,41 @@ public NonUniqueFieldIndex(IndexDefinition<T> indexDefinition) {
   public void add(T object) {
     Object fieldValue = mIndexDefinition.getFieldValue(object);
 
-    ConcurrentHashSet<T> objSet = mIndexMap.get(fieldValue);
+    ConcurrentHashSet<T> objSet;
 
-    // If there is no object set for the current value, creates a new one.
-    if (objSet == null) {
-      mIndexMap.putIfAbsent(fieldValue, new ConcurrentHashSet<T>());
+    do {
```

Comment:
For this block of code, is this equivalent to:

```
do {
  set = <get value from map> (put if absent)
  synchronized (set) {
    if (set != <value in map>) continue;
    set.add(object);
    break;
  }
}
```

This way, we only have to add once to the correct set, and we do the check while holding the lock. Is it still correct if we check without holding the set lock?